core/src/main/java/hudson/diagnosis/OldDataMonitor.java

```diff
@@ -218,32 +215,49 @@ public static void report(Saveable obj, Collection<Throwable> errors) {
             return;
         }
         OldDataMonitor odm = get(j);
-        synchronized (odm) {
-            SaveableReference ref = referTo(obj);
+        SaveableReference ref = referTo(obj);
+        while (true) {
             VersionRange vr = odm.data.get(ref);
-            if (vr != null) vr.extra = buf.toString();
-            else            odm.data.put(ref, new VersionRange(null, buf.toString()));
+            if (vr != null && odm.data.replace(ref, vr, new VersionRange(vr, null, buf.toString()))) {
+                break;
+            } else if (odm.data.putIfAbsent(ref, new VersionRange(null, null, buf.toString())) == null) {
+                break;
+            }
         }
     }
 
     public static class VersionRange {
         private static VersionNumber currentVersion = Jenkins.getVersion();
 
-        VersionNumber min, max;
-        boolean single = true;
-        public String extra;
-
-        public VersionRange(String version, String extra) {
-            min = max = version != null ? new VersionNumber(version) : null;
-            this.extra = extra;
-        }
-
-        public void add(String version) {
-            VersionNumber ver = new VersionNumber(version);
-            if (min==null) { min = max = ver; }
-            else {
-                if (ver.isOlderThan(min)) { min = ver; single = false; }
-                if (ver.isNewerThan(max)) { max = ver; single = false; }
+        final VersionNumber min;
+        final VersionNumber max;
+        final boolean single;
+        final public String extra;
+
+        public VersionRange(VersionRange previous, String version, String extra) {
```

Comment:
I will probably have to refactor in the future to have a real data class that produces the crappy one with public fields for backwards compat.